File: services/benchmark-runner/src/ipw/benchmark_runner/canonical.py
"""Canonical JSON serialisation.

An identity digest is only reproducible if the bytes fed to the hash are
reproducible. This module implements a deliberately **restricted** subset of RFC
8785 (JSON Canonicalization Scheme), chosen so that an equivalent implementation
in JavaScript - which the POC-005 browser lab will need - is a few lines rather
than a port.

Rules
-----
1. Permitted value types: ``dict``, ``list``/``tuple``, ``str``, ``int``,
   ``bool``, ``None``. **Floats are rejected**, not coerced. Float formatting
   differs across platforms and languages and is the single largest source of
   digest drift.
2. Object keys must be non-empty **ASCII** strings. Python sorts by code point
   and JavaScript sorts by UTF-16 code unit; for ASCII the two orders are
   identical, so restricting the alphabet removes the difference by construction.
3. Strings are NFC-normalised.
4. Integers must be exactly representable in IEEE-754 double precision
   (``|n| <= 2**53 - 1``) so JavaScript can reproduce them.
5. Output is compact (no whitespace), UTF-8, with no trailing newline.
"""

from __future__ import annotations

import json
import unicodedata
from collections.abc import Mapping, Sequence
from typing import Any

from ipw.contracts.common import SAFE_INT_MAX, SAFE_INT_MIN

__all__ = ["CanonicalisationError", "canonical_json", "canonical_text", "normalise"]


class CanonicalisationError(TypeError):
    """A value cannot be canonicalised deterministically."""


def _fail(path: str, message: str) -> CanonicalisationError:
    return CanonicalisationError(f"{path or '<root>'}: {message}")


def normalise(value: Any, path: str = "") -> Any:
    """Recursively validate and normalise ``value`` into canonical-safe form."""
    # bool must be tested before int: bool is a subclass of int in Python.
    if value is None or isinstance(value, bool):
        return value

    if isinstance(value, float):
        msg = (
            "float values are forbidden in canonical documents; use an integer "
            "(nanoseconds, bytes, percent) or a decimal string"
        )
        raise _fail(path, msg)

    if isinstance(value, int):
        if not (SAFE_INT_MIN <= value <= SAFE_INT_MAX):
            msg = f"integer {value} is outside the exactly-representable range"
            raise _fail(path, msg)
        return value

    if isinstance(value, str):
        return unicodedata.normalize("NFC", value)

    if isinstance(value, Mapping):
        out: dict[str, Any] = {}
        for raw_key, raw_value in value.items():
            if not isinstance(raw_key, str):
                raise _fail(path, f"object keys must be strings, found {type(raw_key).__name__}")
            if not raw_key:
                raise _fail(path, "object keys must be non-empty")
            if not raw_key.isascii():
                msg = (
                    f"object key {raw_key!r} must be ASCII so that key ordering is "
                    "identical in Python and JavaScript"
                )
                raise _fail(path, msg)
            if raw_key in out:
                raise _fail(path, f"duplicate object key {raw_key!r}")
            out[raw_key] = normalise(raw_value, f"{path}/{raw_key}")
        return out

    if isinstance(value, Sequence) and not isinstance(value, str | bytes | bytearray):
        return [normalise(item, f"{path}/{index}") for index, item in enumerate(value)]

    raise _fail(path, f"unsupported type {type(value).__name__}")


def canonical_text(value: Any) -> str:
    """Return the canonical JSON text for ``value``."""
    return json.dumps(
        normalise(value),
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,
    )


def canonical_json(value: Any) -> bytes:
    """Return the canonical UTF-8 bytes for ``value``. These are what get hashed."""
    return canonical_text(value).encode("utf-8")

File: packages/contracts-ts/src/canonical.ts
/**
 * Canonical JSON serialisation - the TypeScript half of the contract.
 *
 * This must produce byte-identical output to
 * `services/benchmark-runner/src/ipw/benchmark_runner/canonical.py`. If the two
 * ever disagree, a digest computed in the browser will not match one computed by
 * the runner, and results that are actually the same will look different.
 *
 * Two POC-001 decisions exist specifically to make that agreement achievable, and
 * this file is where they pay off:
 *
 * - **No floats.** JavaScript has one number type, so `1.0` and `1` are the same
 *   value and serialise identically. Python distinguishes them and rejects the
 *   float. The rule is enforced on both sides as "integers only", which lands on
 *   the same bytes either way.
 * - **ASCII-only object keys.** Python sorts strings by code point;
 *   `Array.prototype.sort` sorts by UTF-16 code unit. Those orders differ above
 *   U+FFFF. Restricting keys to ASCII makes them identical by construction, so
 *   neither side needs a custom comparator.
 */

/** JavaScript's `Number` is exact only to 2^53-1; Python's ints are not. */
export const SAFE_INT_MAX = Number.MAX_SAFE_INTEGER;
export const SAFE_INT_MIN = Number.MIN_SAFE_INTEGER;

/** A value that can appear in a canonical document. */
export type CanonicalValue =
  | string
  | number
  | boolean
  | null
  | CanonicalValue[]
  | { [key: string]: CanonicalValue };

export class CanonicalisationError extends Error {
  readonly path: string;

  constructor(path: string, message: string) {
    super(`${path || "<root>"}: ${message}`);
    this.name = "CanonicalisationError";
    this.path = path;
  }
}

const ASCII = /^[\x20-\x7e]+$/;

/**
 * Validate and normalise a value into canonical-safe form.
 *
 * Rejects rather than coerces. A silent coercion here would produce a digest
 * that looks valid and matches nothing.
 */
export function normalise(value: unknown, path = ""): CanonicalValue {
  if (value === null) return null;

  const kind = typeof value;

  if (kind === "boolean") return value as boolean;

  if (kind === "number") {
    const n = value as number;
    if (!Number.isFinite(n)) {
      throw new CanonicalisationError(path, "NaN and Infinity are not representable in JSON");
    }
    if (!Number.isInteger(n)) {
      throw new CanonicalisationError(
        path,
        "non-integer numbers are forbidden in canonical documents; use an integer " +
          "(nanoseconds, bytes, percent) or a decimal string",
      );
    }
    if (n < SAFE_INT_MIN || n > SAFE_INT_MAX) {
      throw new CanonicalisationError(path, `integer ${n} is outside the exactly-representable range`);
    }
    return n;
  }

  if (kind === "string") {
    const s = value as string;
    // A lone surrogate has no valid UTF-8 encoding, so the two languages could
    // not agree on its bytes even in principle.
    if (typeof (s as { isWellFormed?: () => boolean }).isWellFormed === "function" && !s.isWellFormed()) {
      throw new CanonicalisationError(path, "string contains an unpaired surrogate");
    }
    return s.normalize("NFC");
  }

  if (Array.isArray(value)) {
    return value.map((item, index) => normalise(item, `${path}/${index}`));
  }

  if (kind === "object") {
    const source = value as Record<string, unknown>;
    const out: Record<string, CanonicalValue> = {};
    for (const key of Object.keys(source)) {
      if (key.length === 0) {
        throw new CanonicalisationError(path, "object keys must be non-empty");
      }
      if (!ASCII.test(key)) {
        throw new CanonicalisationError(
          path,
          `object key ${JSON.stringify(key)} must be ASCII so that key ordering is ` +
            "identical in Python and JavaScript",
        );
      }
      out[key] = normalise(source[key], `${path}/${key}`);
    }
    return out;
  }

  throw new CanonicalisationError(path, `unsupported type ${kind}`);
}

/**
 * Serialise a normalised value with sorted keys and no whitespace.
 *
 * Written by hand rather than via `JSON.stringify(value, Object.keys(...).sort())`
 * because the replacer form does not sort nested objects reliably, and key order
 * is the whole point.
 */
function serialise(value: CanonicalValue): string {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") return String(value);
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(serialise).join(",")}]`;

  const keys = Object.keys(value).sort();
  const members = keys.map((key) => `${JSON.stringify(key)}:${serialise(value[key] as CanonicalValue)}`);
  return `{${members.join(",")}}`;
}

/** Canonical JSON text. Compact, sorted, NFC, no trailing newline. */
export function canonicalText(value: unknown): string {
  return serialise(normalise(value));
}

/** Canonical UTF-8 bytes. These are what get hashed. */
export function canonicalBytes(value: unknown): Uint8Array {
  return new TextEncoder().encode(canonicalText(value));
}
